src/main/java/spliterators/part1/exercise/RectangleSpliterator.java

```diff
@@ -33,49 +33,57 @@ private RectangleSpliterator(int[][] array, int startOuterInclusive, int endOute
         }
-        if (length == 1){
-            return tryInnerSplit();
-        }
-        final int middle = startOuterInclusive + length /2;
-        RectangleSpliterator result = new RectangleSpliterator(array,startOuterInclusive,middle,startInnerInclusive,
-                endInnerExclusive);
-        startOuterInclusive = middle;
-        return result;
-    }
-
-    private OfInt tryInnerSplit() {
-        final int length = endInnerExclusive - startInnerInclusive;
-        if (length < 2){
+        if (totalLength < 2){
             return null;
         }
-        final int middle = startInnerInclusive + length / 2;
-        RectangleSpliterator result = new RectangleSpliterator(array,startOuterInclusive,endOuterExclusive,
-                startInnerInclusive,middle);
-        startInnerInclusive = middle;
+        final RectangleSpliterator result;
+        int lengthLeft = totalLength / 2;
+        if (lengthLeft - (innerLength - startInnerInclusive) <= 0){     // outerLength == 1
+            result = new RectangleSpliterator(array,startOuterInclusive,startOuterInclusive + 1,startInnerInclusive,
+                    startInnerInclusive + lengthLeft);
+            startInnerInclusive = startInnerInclusive + lengthLeft;
+            return result;
+        }
+        lengthLeft -= (innerLength - startInnerInclusive);
+        final int fullRowsCount = Math.floorDiv(lengthLeft,innerLength); // outerLength > 1
+        lengthLeft -= fullRowsCount * innerLength;
+        result = new RectangleSpliterator(array,startOuterInclusive,startOuterInclusive + 2 + fullRowsCount,startInnerInclusive,
+                lengthLeft);
+        startInnerInclusive = lengthLeft;
+        startOuterInclusive += 1 + fullRowsCount;
         return result;
     }
 
     @Override
     public long estimateSize() {
-        return ((long) endOuterExclusive - startOuterInclusive)*(endInnerExclusive - startInnerInclusive);
+        int outerLength = endOuterExclusive - startOuterInclusive;
+        if (outerLength == 1){
+            return endInnerExclusive - startInnerInclusive;
+        } else{
+            return (innerLength - startInnerInclusive) + innerLength * (outerLength - 2) + endInnerExclusive;
+        }
     }
 
     @Override
     public boolean tryAdvance(IntConsumer action) {
+        if (startInnerInclusive >= innerLength){
+            startInnerInclusive = 0;
+            ++startOuterInclusive;
+        }
         if (startOuterInclusive >= endOuterExclusive){
             return false;
         }
-        if (startInnerInclusive >= endInnerExclusive) {
-            ++startOuterInclusive;
-            if (startOuterInclusive >= endOuterExclusive){
-                return false;
-            }
-            startInnerInclusive = 0;
-            endInnerExclusive = array[startOuterInclusive].length;
+        if (endOuterExclusive == startOuterInclusive + 1 && startInnerInclusive >= endInnerExclusive) {

```

Comment:
Your solution can cause `action.accept` with `startInnerInclusive == endInnerExclusive`
Is it allowed?
Moreover `startInnerInclusive = 0;` isn't executed everytime